### PR TITLE
ENH: stats: add `alternative` to masked normality tests

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

Addresses gh-12506.
gh-13549 added the alternative parameter to some normality tests but forgot to add it to the masked version. This is a continuation of that work.

#### What does this implement/fix?

`skewtest` and `kurtosistest` were missing an `alternative` parameter
in their masked version. It has been added and tested now.

#### Additional information

N/A

### CODE REVIEW

File: scipy/stats/mstats_basic.py
Comment (tylerjereddy) [inline scipy/stats/mstats_basic.py:2553]: Looks like codecov is saying the error handling code isn't tested here.

File: scipy/stats/mstats_basic.py
Comment (tylerjereddy) [inline scipy/stats/mstats_basic.py:2614]: same here

File: scipy/stats/mstats_basic.py
Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:2553]: Oops, forgot to test the error! Now, done. Thanks.

File: scipy/stats/mstats_basic.py
Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:2614]: Done.

File: scipy/stats/tests/test_mstats_basic.py
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:966]: This currently compares the output of these functions against the `stats` versions when there are no masked values. Can you test it with masked values, and for the `stats` version use the [`compressed`](https://numpy.org/doc/stable/reference/generated/numpy.ma.MaskedArray.compressed.html) array?

Even better would be to test the multi-dimensional case, parametrizing over the values of the `axis` argument. But for that, we can wait until gh-13312 gets merged. Then it will be really easy to add multi-dimensional tests for a `mstats` and `nan_policy='omit'` cases.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4657]: These are OK if we assume that `alternative='two-sided'` with `nan_policy='omit'` is well-tested. Do you think that is the case?

**Review by mdhaber — COMMENTED** [review]
Good start, but please review what we did in gh-13549 and apply some of that here. For instance, better document the meaning of the `less` vs `greater` and rely on `_normtest_finish` to raise the error for an invalid `alternative` argument.

File: scipy/stats/tests/test_mstats_basic.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_mstats_basic.py:966]: > Can you test it with masked values, and for the `stats` version use the [`compressed`](https://numpy.org/doc/stable/reference/generated/numpy.ma.MaskedArray.compressed.html) array?

Sure!



> Even better would be to test the multi-dimensional case, parametrizing over the values of the `axis` argument. But for that, we can wait until [gh-13312](https://github.com/scipy/scipy/pull/13312) gets merged.

That's a great idea too. I can wait until gh-13312 gets merged. I will keep this in mind until then!

**Comment by tirthasheshpatel** [comment]
Hi, @mdhaber. Thanks for the review!

> better document the meaning of the `less` vs `greater`

Ah, I believe I didn't document the masked versions as its documentation points to the `stats` version where `less` and `greater` are documented in more detail. But I am also OK with copy-pasting those docs to the masked versions. What do you think?

> rely on `_normtest_finish` to raise the error for an invalid `alternative` argument

On an unrelated note, relying on `_normtest_finish` would result in computing the Z value and then throwing the error. What we probably want is to throw an error before doing all the heavy computations. I don't know if it is a big problem but just something I thought would be worth pointing out/discussing.

File: scipy/stats/tests/test_stats.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_stats.py:4657]: > Do you think that is the case?

Yes, I see a test for that case for both `skewtest` and `kurtosistest`. I can remove these then if you want.

**Comment by mdhaber** [comment]
> But I am also OK with copy-pasting those docs to the masked versions. What do you think?

I do think that the definitions should appear here; we should not just refer the user to `stats` for documentation. (If we were to refer the user to `stats` for a portion of the documentation, why not refer the user to `stats` for all of the function documentation?)

For implementing this, it would be nice to define the text (e.g. as a variable) in one place and use it in many places, but for now let's copy-paste. There are a lot of other places where we share text that could be cleaned up at the same time (if we decide to go that route at some point).

> On an unrelated note, relying on `_normtest_finish` would result in computing the Z value and then throwing the error

I agree it's not ideal, but there is the same problem in the `stats` version. I'd keep it consistent here. Maybe just rely on `_normtest_finish` here, and if you are so inspired, have another PR that does early input validation for hypothesis tests (and remove it from `_normtest_finish`).

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:4657]: No, I don't mean to remove these. I'm saying that the test is only valid if we trust that `mstats.skewtest` with `nan_policy='omit'` does the right thing. 

I just don't think that the existing test of `mstats.skewtest` with `nan_policy='omit'` is sufficient. But I don't think that is something you have to fix in this PR, so never mind.

**Comment by tirthasheshpatel** [comment]
> I do think that the definitions should appear here

Got it. copy-pasted the docs in the latest commit.

> I agree it's not ideal, but there is the same problem in the `stats` version. I'd keep it consistent here.

Makes sense, done.

> if you are so inspired, have another PR that does early input validation for hypothesis tests (and remove it from `_normtest_finish`).

:+1:

File: scipy/stats/tests/test_mstats_basic.py
Comment (mdhaber) [inline scipy/stats/tests/test_mstats_basic.py:967]: This is fine - as long as it wasn't necessary to make the test pass...? What was the motivation?

**Review by mdhaber — APPROVED** [review]

File: scipy/stats/mstats_basic.py
Comment (mdhaber) [inline scipy/stats/mstats_basic.py:2605]: More accurately, these would be "scalar or ndarray".

Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:2605]: > More accurately, these would be "scalar or ndarray". (Could be scalar, and if it's array-like, it's going to be an ndarray)

I think `array_like` includes both scalars and array outputs. So, we can change it that way.

File: scipy/stats/tests/test_mstats_basic.py
Comment (tirthasheshpatel) [reply scipy/stats/tests/test_mstats_basic.py:967]: Oh, yes. It wasn't done to make the test pass but because `skewtest` requires at least 20 samples. As I add nans to some samples, I had to increase the sample size.

File: scipy/stats/mstats_basic.py
Comment (mdhaber) [reply scipy/stats/mstats_basic.py:2605]: I'm not sure if `array_like` is formally defined anywhere, so I guess we can use it to mean whatever we want. But we could be more specific as we push toward better documentation.

Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:2605]: > I'm not sure if `array_like` is formally defined anywhere, so I guess we can use it to mean whatever we want. But we could be more specific as we push toward better documentation. (Not required for this PR, though.)

I think that NumPy defines `array_like` in https://numpydoc.readthedocs.io/en/latest/format.html#other-points-to-keep-in-mind . See gh-13621

Comment (mdhaber) [reply scipy/stats/mstats_basic.py:2605]: Oh, nice. Thanks for pointing that out. Note that it only specifically says it's for documenting _arguments_, though. We are flexible about arguments, but we know what the return types will be. 

But even e.g. `np.mean` doesn't document this perfectly. It says that the output type will be ndarray
image

but as we know:
```
isinstance(np.mean([[1, 2, 3]], np.ndarray)  # True
isinstance(np.mean([1, 2, 3]), np.ndarray)  # False
```
I'll go ahead and merge this as-is, if it sounds good to you? 
We can consider adding this to the huge list of things we'd like the documentation to be more consistent about.

Comment (tirthasheshpatel) [reply scipy/stats/mstats_basic.py:2605]: > if it sounds good to you?

Yep, everything sounds good to me!

**Comment by mdhaber** [comment]
This PR adds support for an `alternative` argument to the `mstats` versions of `skewtest` and `kurtosistest`; it also enables the `stats` versions of these tests to use `alternative` and `nan_policy='omit'`. The behavior is what we'd get if we were to manually remove `nan`s from the arrays, e.g.:
```python3
import numpy as np
from scipy import stats
from scipy.stats import mstats

np.random.seed(0)

for test_name in {"kurtosistest", "skewtest"}:
    for alternative in {'less', 'greater', 'two-sided'}:
        for i in range(100):
            sample = stats.norm.rvs(size=np.random.randint(30, i+40))
            p = np.random.rand()
            mask = np.random.rand(*sample.shape) > (0.5 + p/2)
            
            test = getattr(stats, test_name)
            mtest = getattr(mstats, test_name)

            compressed_sample = sample[~mask]

            nan_sample = sample.copy()
            nan_sample[mask] = np.nan

            masked_sample = np.ma.masked_array(sample, mask=mask)

            res1 = test(compressed_sample)
            res2 = test(nan_sample, nan_policy='omit')
            res3 = mtest(masked_sample)

            np.testing.assert_allclose(res2, res1)
            np.testing.assert_allclose(res3, res1)
```
Tests are fine (considering the existing tests of `skewtest` and `kurtosistest`) and documentation of the `alternative`s is thorough. 

Thanks @tirthasheshpatel!

**Pull request merged**
